### ISSUE TITLE
Failed to Save Kubernetes certificates: Timeout waiting for K8s to be ready

### ISSUE THREAD

**Issue by HighwayofLife** [rancher/rke#121]
**RKE version:**
rke version v0.0.8-dev

**Docker version: (`docker version`,`docker info` preferred)**
Docker version 17.03.2-ce, build f5ec1e2

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
Azure (UbuntuServer VM, no extensions)

**cluster.yml file:**
```yml
nodes:
- address: 10.18.160.13
  role:
  - controlplane
  - etcd
  hostname_override: kubemaster
  user: <redacted>
- address: 10.18.160.14
  role:
  - worker
  hostname_override: kubenode1
  user: <redacted>
- address: 10.18.160.15
  role:
  - worker
  hostname_override: kubenode2
  user: <redacted>
services:
  etcd:
    image: quay.io/coreos/etcd:latest
  kube-api:
    image: rancher/k8s:v1.8.3-rancher2
    service_cluster_ip_range: 10.234.0.0/18
  kube-controller:
    image: rancher/k8s:v1.8.3-rancher2
    cluster_cidr: 10.233.64.0/18
    service_cluster_ip_range: 10.234.0.0/18
  scheduler:
    image: rancher/k8s:v1.8.3-rancher2
  kubelet:
    image: rancher/k8s:v1.8.3-rancher2
    cluster_domain: cluster.local
    infra_container_image: gcr.io/google_containers/pause-amd64:3.0
    cluster_dns_server: 10.234.0.3
  kubeproxy:
    image: rancher/k8s:v1.8.3-rancher2
network:
  plugin: flannel
auth:
  strategy: x509
ssh_key_path: ~/.ssh/kubernetes
```

**Steps to Reproduce:**
```./rke --debug up```

**Results:**
```
...
INFO[2597] [worker] Successfully started [kube-proxy] container on host [10.18.160.18] 
INFO[2597] [worker] Successfully started Worker Plane.. 
INFO[2597] [certificates] Save kubernetes certificates as secrets 
DEBU[2597] [certificates] Saving certificate [kube-scheduler] to kubernetes 
FATA[2627] [certificates] Failed to Save Kubernetes certificates: Failed to save certificate [kube-scheduler] to kubernetes: [certificates] Timeout waiting for kubernetes to be ready 
```

Tried several times on 2 different clusters configured the same way. It always fails at this point. Saving certificates to kubernetes.

**Comment by HighwayofLife** [comment]
A bit more detail. Once the VMs are newly provisioned, I install docker using this script.
Following that is when I run rke against these machines. No other configuration is being done on them.

```sh
#!/bin/sh

sudo apt-get update
sudo apt-get install -y \
    apt-transport-https \
    ca-certificates \
    curl \
    software-properties-common

 curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
 sudo add-apt-repository \
   "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
   $(lsb_release -cs) \
   stable"

 sudo apt-get update
 sudo apt-get install -y docker-ce=17.03.2~ce-0~ubuntu-xenial
 sudo docker run hello-world
 sudo usermod -aG docker $USER
 sudo systemctl enable docker
```

**Comment by mmuruganandam** [comment]
I have the similar problem.

**Comment by superseb** [comment]
The kubernetes components are setup using SSH, after setting it up the certificates are saved to kubernetes using the setup certificates to the apiserver on port 6443. My guess would be that port 6443 is being blocked in your env/on your host. Could be nice to do this over the SSH tunnel as well, to minimize failures like this. Will discuss, in the meantime, can you try opening up the port and see if that works?

**Comment by xrain0610** [comment]
The same problem.  the port 6443 is open. i think ,is the too short timeout ?

**Comment by galal-hussein** [comment]
i don't think its a short timeout, can you check the docker containers on the machines and make sure nothing is in restarting loop, also what is the output of `kubectl describe nodes
`

**Comment by HighwayofLife** [comment]
I thought it might be due to 6443 being blocked to inbound traffic. It is, but I can tunnel into one of the servers, and it still will fail.
From my terminal to the rke host (master), and tested from one of the worker nodes (they are all on the same subnet):
```telnet 10.18.160.14 6443
Trying 10.18.160.14...
Connected to 10.18.160.14.
Escape character is '^]'.
^]
telnet> quit
Connection closed.```

The API server was in a restart loop the first time I looked.

**Comment by galal-hussein** [comment]
@HighwayofLife can you send kube-api logs and kubelet as well

**Comment by galal-hussein** [comment]
@HighwayofLife sorry i don't understand what you mean by ` I created an SSH tunnel FIRST (not after it failed)`, how do you create the ssh tunnel?

**Comment by HighwayofLife** [comment]
1. Spin up Virtual Machines in Azure cloud
2. Install Docker
3. Run rke against machines
-- Fails:
```
FATA[2627] [certificates] Failed to Save Kubernetes certificates:
  Failed to save certificate [kube-scheduler] to kubernetes: [certificates]
  Timeout waiting for kubernetes to be ready 
```
4. SSH Tunnel into one of the VMs
5. Run rke again against machines
-- Still fails.
I had noticed that the API Server seems to be in a restart loop in this case.

### Start Over...
1. Spin up Virtual Machines in Azure cloud
2. SSH Tunnel into one of the VMs (does not matter which one)
```sh
sshuttle -r <username>@10.18.160.13  10.18.160.0/24
```
3. This will direct all traffic in the 10.18.160.0/24 subnet through the SSH Tunnel on that VM. Effectively bypassing the port 6443 that's being blocked by our network layer.
4. Install Docker
5. Run rke against machines.
-- SUCCESS!

I just created 2 clusters successfully using this method. Yesterday, I failed at least 20+ times to create a single cluster.

To summarize, there is an issue where the target servers appear to require certain network access to the host running rke. It's only a guess what that would be, assumption is that 6443 needs to be open. -- It shouldn't need this since it should be relying on the tunnel. If it does need it, then it just needs to be listed in the prerequisites.

@xrain0610 and @mmuruganandam do you guys want to try the steps with sshuttle into one of the target servers prior to running rke and see if it works for you?

**Comment by HighwayofLife** [comment]
Without the tunnel. I searched for logs, but couldn't find any.
`sudo journalctl -r -u kubelet` shows `-- no entries --`

```sh
$ kubectl get nodes
NAME          STATUS     ROLES     AGE       VERSION
kubenode1     NotReady   worker    3m        v1.8.3-rancher1
kubenode2     NotReady   worker    1m        v1.8.3-rancher1
kubeserver1   NotReady   master    4m        v1.8.3-rancher1
```

```sh
$ docker ps
CONTAINER ID        IMAGE                         COMMAND                  CREATED             STATUS              PORTS               NAMES
efcf838d4972        rancher/k8s:v1.8.3-rancher2   "kube-proxy --v=2 ..."   13 minutes ago      Up 13 minutes                           kube-proxy
e3323638a3a3        rancher/k8s:v1.8.3-rancher2   "kubelet --v=2 --a..."   13 minutes ago      Up 13 minutes                           kubelet
9e01de920295        rancher/k8s:v1.8.3-rancher2   "kube-scheduler --..."   13 minutes ago      Up 13 minutes                           scheduler
8ad5f86263b6        rancher/k8s:v1.8.3-rancher2   "kube-controller-m..."   13 minutes ago      Up 13 minutes                           kube-controller
0464f96f2a77        rancher/k8s:v1.8.3-rancher2   "kube-apiserver --..."   13 minutes ago      Up 13 minutes                           kube-api
```

**Comment by HighwayofLife** [comment]
And one final note for anybody interested later down the road.
If you run into this problem, then create the tunnel, you can run `rke remove` to remove the cluster, and then (while tunneled) you can do `rke up` to create the cluster without recreating/wiping your machines.

**Comment by luoshiqian** [comment]
i have the same problem,when run "rke -d up".the error show me is
FATA[2627] [certificates] Failed to Save Kubernetes certificates:
  Failed to save certificate [kube-proxy] to kubernetes: [certificates]
  Timeout waiting for kubernetes to be ready.
and i resolve it by run "rke remove" and then run "rke -d up" again

**Comment by galal-hussein** [comment]
@HighwayofLife Kubernetes api is running on port 6443 in rke, if this port is opened between hosts in your private network then rke should start successfully, are you still seeing the problem or its save to close the issue

**Comment by HighwayofLife** [comment]
@galal-hussein 
> Kubernetes api is running on port 6443 in rke
If that's the case, then there are multiple problems (bugs?)

1. `rke` and rke docs do not list port 6443 being open as a prerequisites.
2. Why the need for an SSH Tunnel if `rke` isn't using it?
3. Why isn't `rke` testing port 6443 to the hosts to ensure that the port is open?
4. `rke` starts just fine, but kubernetes is not becoming fully ready on the nodes. Unclear why because there is no error handling around the actual failure.

**Comment by galal-hussein** [comment]
@HighwayofLife 
1. We are still working on the documentation for RKE that will include more details, there is already opened PR for more detailed README

2. User shouldn't worry about ssh tunneling at all, RKE will use ssh tunnel connection as a custom dialer for Docker connection https://github.com/rancher/rke/blob/master/hosts/dialer.go#L33-L53, so the only requirement for the user to have ssh connection available for the hosts.

3. RKE is a kubernetes installer, it doesn't work as a monitoring tool however rke will fail if it can't connect to the 6443 port, and it will fail to save the certificates in k8s

4. can you provider more logs about this error like:
```
kubectl describe nodes
kubectl describe pods -n kube-system
docker logs kube-api
docker logs kubelet
```

**Comment by mmuruganandam** [comment]
Opening 6443 makes the deployment go through without issues

**Comment by greatbn** [comment]
I have same error. Port 6443 is opened.

**Comment by Kamaradeivanov** [comment]
Hi, I got the same problem on GCE. But I solved it with the help of @superseb.
All my ports are opened between private network interfaces. 
But only port 22 are open on public interfaces. 

I just open port 6443 on public network and everything works well know. BTW I has no other choice to open it to use kubectl from my post to contact my k8s cluster.

**Comment by HighwayofLife** [comment]
The `kube-apiserver --secure-port` default is 6443, but this is configurable, and would be nice to be able to configure within rke as an option. Many clusters are created with other tools that are created with secure port 443 (for example).

**Comment by chanwit** [comment]
Even my 6443 port in the security group is opened, I still got this error and the cluster is failed to be provisioned.

The environment is AWS.

**Comment by galal-hussein** [comment]
The error is considered a general error that means that kubernetes cluster is not running probably hence saving certificates to k8s will fail, so it may mean that the port is not opened or kube-api can't connect to etcd, etc.

we opened an issue for rke to validate components after deployment #152

**Comment by galal-hussein** [comment]
@HighwayofLife Closing this issue as #152 and #212 should be able to validate healthchecks for services and make sure the right ports opened on each host.
